odoo/tools/safe_eval.py

```diff
@@ -1,225 +1,812 @@
+        """Function that checks if a given object is safe.
 
-    # set operations are almost twice as fast as a manual iteration + condition
-    # when loading /web according to line_profiler
-    code_codes = {i.opcode for i in dis.get_instructions(code_obj)}
-    if not allowed_codes >= code_codes:
-        raise ValueError("forbidden opcode(s) in %r: %s" % (expr, ', '.join(opname[x] for x in (code_codes - allowed_codes))))
+        This function is used to check if a given object is safe to be used inside of the evaluation context.
 
-    for const in code_obj.co_consts:
-        if isinstance(const, CodeType):
-            assert_valid_codeobj(allowed_codes, const, 'lambda')
+        A safe object is an object that is either:
+            - A built-in object (See _BUILTINS)
+            - An allowed type (See _ALLOWED_TYPES)
+            - An allowed instance (See _ALLOWED_INSTANCES)
+            - A wrapped object (See SafeWrapper)
+            - A wrapped type (See _wrapped_types)
+            - A wrapped instance (See _wrapped_instances)
+            - A wrapped module (See _ALLOWED_MODULES)
 
-def test_expr(expr, allowed_codes, mode="eval", filename=None):
-    """test_expr(expression, allowed_codes[, mode[, filename]]) -> code_object
+        We uses a white-list approach but with some exceptions that are checked in the function.:
+            - The safe_eval function
+            - Objects that their __self__ is an unsafe module (We can prevent functions of the like of os.system that way)
+                Note : Objects present in _BUILTINS are not checked, as they are already safe
+
+        Raises:
+            ValueError: If the object is not safe
+
+        Args:
+            __obj (Any): The object to be checked
+        """
 
-    Test that the expression contains only the allowed opcodes.
-    If the expression is valid and contains only allowed codes,
-    return the compiled code object.
-    Otherwise raise a ValueError, a Syntax Error or TypeError accordingly.
+        obj = args.pop("__obj")
+
+        # The black-list part (See docstring)
+        if obj is safe_eval or (
+            callable(obj)
+            and not isinstance(obj, SafeWrapper)
+            and hasattr(obj, "__self__")
+            and isinstance(obj.__self__, types.ModuleType)
+            and obj.__self__.__name__ not in _ALLOWED_MODULES
+            and obj not in _BUILTINS.values()
+        ):
+            raise ValueError(f"unsafe object: {obj} (denied by black-list)")
+
+        if isinstance(obj, allowed_instance) or (
+            isinstance(obj, type) and issubclass(obj, allowed_type)
+        ):
+            return obj
+        elif (
+            isinstance(obj, _wrapped_instances)
+            or (isinstance(obj, type) and issubclass(obj, _wrapped_types))
+            or (isinstance(obj, types.ModuleType) and obj.__name__ in _ALLOWED_MODULES)
+        ):
+            return SafeWrapper(__obj=obj, __type_checker=type_checker)
```

Comment:
if I do that change, all the tests are green (so either it's useles, either some tests are missing)
```suggestion
            return obj